File: leetcode-cn/0899.0_Orderly_Queue.py
'''
1. k == 1 时，只能轮转；
2. k > 1 时，可以实现任意两个元素交换；遂可实现冒泡排序。

e.g.:  k == 2
   c d e f g h
-> c [e f g h] d
-> [e f g h] d c
-> d c [e f g h] 跟最初比较，只是交换了c d.

执行用时：44 ms, 在所有 Python3 提交中击败了21.43% 的用户
内存消耗：14.9 MB, 在所有 Python3 提交中击败了79.46% 的用户
通过测试用例：110 / 110
'''
class Solution:
    def orderlyQueue(self, s: str, k: int) -> str:
        if k > 1:
            return ''.join(sorted(s))
        ans = s
        for i in range(len(s)):
            ans = min(ans, s[i: ] + s[: i])
        return ans 

File: csharp/0899.0_Orderly_Queue.cs
/*
Runtime: 148 ms, faster than 14.29% of C# online submissions for Orderly Queue.
Memory Usage: 39.1 MB, less than 14.29% of C# online submissions for Orderly Queue.
*/
public class Solution {
    public string OrderlyQueue(string s, int k) {
        if (k > 1) {
            return string.Concat(s.OrderBy(x => x));
        }
        string ans = s;
        foreach (char c in s) {
            s = s[1..] + s[0];
            if (s.CompareTo(ans) < 0) ans = s;
        }
        return ans;
    }
}


/*
Runtime: 150 ms, faster than 14.29% of C# online submissions for Orderly Queue.
Memory Usage: 39 MB, less than 14.29% of C# online submissions for Orderly Queue.
*/
public class Solution {
    public string OrderlyQueue(string s, int k) => k > 1 ?
        string.Concat(s.OrderBy(x => x)) :
        s.Min(c => s = $"{s[1..]}{c}");
}


/*
Runtime: 162 ms, faster than 14.29% of C# online submissions for Orderly Queue.
Memory Usage: 38.6 MB, less than 14.29% of C# online submissions for Orderly Queue.
*/
public class Solution {
    public string OrderlyQueue(string s, int k) {
        if (1 == k) {
            string ans = s;
            for (int i = 1; i < s.Length; ++i) {
                string m = s[i..] + s[..i];
                if (m.CompareTo(ans) < 0) ans = m;
            }
            return ans;
        }
        char[] characters = s.ToCharArray();
        Array.Sort(characters);
        return new string(characters);
    }
}

/*
Runtime: 140 ms, faster than 28.57% of C# online submissions for Orderly Queue.
Memory Usage: 38.7 MB, less than 14.29% of C# online submissions for Orderly Queue.
*/
public class Solution {
    public string OrderlyQueue(string s, int k) {
        if (1 == k) {
            string sTwice = s + s;
            for (int i = 1; i < s.Length; ++i) {
                string m = sTwice.Substring(i, s.Length);
                if (m.CompareTo(s) < 0) s = m;
            }
            return s;
        }
        char[] characters = s.ToCharArray();
        Array.Sort(characters);
        return new string(characters);        
    }
}



/*
Runtime: 163 ms, faster than 14.29% of C# online submissions for Orderly Queue.
Memory Usage: 39.1 MB, less than 14.29% of C# online submissions for Orderly Queue.
*/
public class Solution {
    public string OrderlyQueue(string s, int k) 
        => k > 1
            ? string.Join("", s.OrderBy(x => x))
            : Enumerable.Range(0, s.Length).Select(i => s[i..] + s[..i]).Min()!;
}
